Framework/Algorithms/src/MaskNonOverlappingBins.cpp

```diff
@@ -0,0 +1,136 @@
+std::string const OUTPUT_WS{"OutputWorkspace"};
+} // namespace Prop
+} // namespace
+
+namespace Mantid {
+namespace Algorithms {
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+DECLARE_ALGORITHM(MaskNonOverlappingBins)
+
+/// Algorithms name for identification. @see Algorithm::name
+std::string const MaskNonOverlappingBins::name() const {
+  return "MaskNonOverlappingBins";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int MaskNonOverlappingBins::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+std::string const MaskNonOverlappingBins::category() const {
+  return "Transforms\\Masking";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+std::string const MaskNonOverlappingBins::summary() const {
+  return "Marks bins in InputWorkspace which are out of the X range of the "
+         "second workspace.";
+}
+
+/// Return a list of the names of related algorithms.
+std::vector<std::string> const MaskNonOverlappingBins::seeAlso() const {
+  return {"MaskBins", "MaskBinsIf"};
+}
+
+/** Initialize the algorithm's properties.
+ */
+void MaskNonOverlappingBins::init() {
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::INPUT_WS, "", Kernel::Direction::Input),
+                  "A workspace to mask.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::OUTPUT_WS, "", Kernel::Direction::Output),
+                  "The masked workspace.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::COMPARISON_WS, "", Kernel::Direction::Input),
+                  "A workspace to compare the InputWorkspace's binning to.");
+  declareProperty(Prop::MASK_PARTIAL, false,
+                  "If true, mask also bins that overlap only partially.");
+}
+
+/// Returns a map from property name to message in case of invalid values.
+std::map<std::string, std::string> MaskNonOverlappingBins::validateInputs() {
+  std::map<std::string, std::string> issues;
+  API::MatrixWorkspace_const_sptr inputWS = getProperty(Prop::INPUT_WS);
+  API::MatrixWorkspace_const_sptr comparisonWS =
+      getProperty(Prop::COMPARISON_WS);
+  if (inputWS->getNumberHistograms() != comparisonWS->getNumberHistograms()) {
+    issues[Prop::COMPARISON_WS] =
+        "The number of histogams mismatches with " + Prop::INPUT_WS;
+  }
+  if (!inputWS->isHistogramData()) {
+    issues[Prop::INPUT_WS] =
+        "The workspace contains point data, not histograms.";
+  }
+  if (!comparisonWS->isHistogramData()) {
+    issues[Prop::COMPARISON_WS] =
+        "The workspace contains point data, not histograms.";
+  }
```

Comment:
Should you also check the unit type. For example I think the current method would allow you to provide workspaces with different x-units? Further, your solution relies on the X-axis being sorted. This is usually the case, but we've had issues in the past with user-defined data not being sorted correctly (hence `SortXAxis` algorithm). Also It's possible for `EventWorkspaces` to have a sorting other than TOF - there is for example a `RebinByPulseTimes` algorithm.

Reply:
X units checking sound like a neat idea. I'll have a look at the X sorting stuff as well.